### DISCUSSION TITLE
Multiple ORs failing

### DISCUSSION THREAD

**Discussion by JasminDreasond** [ujjwalguptaofficial/JsStore#387] [Q&A] [answered]
I'm trying to make something like it:
```sql
Select * from Table_Name where Column1=value1 and (Column2=value2 or Column2=value3 or Column2=value4);
```
```js
if(Column1 === value1 && (Column2 === value2 || Column2 === value3 || Column2 === value4))
  return true;
```
EDIT: I have a value of a message to receive that already has the values of which Room belongs, but I'm adding another condition that this value also needs to vary between 3 responses of the same string input.

But that always fails, returning an array without my result

reference to test:
https://ujjwalguptaofficial.github.io/idbstudio/?db=Demo&query=select(%7B%0A%20%20%20%20from%3A%20%22Customers%22%2C%0A%20%20%20%20where%3A%20%5B%7B%0A%20%20%20%20%20%20%20%20or%3A%20%5B%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20country%3A%20'USA'%0A%20%20%20%20%20%20%20%20%7D%2C%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20country%3A%20'Mexico'%0A%20%20%20%20%20%20%20%20%7D%2C%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20country%3A%20'Brazil'%0A%20%20%20%20%20%20%20%20%7D%5D%0A%20%20%20%20%7D%5D%0A%7D)%3B

**Comment by ujjwalguptaofficial** [comment]
It's giving me results - 

<img width="1440" alt="Screenshot 2024-11-29 at 1 17 36 PM" src="https://github.com/user-attachments/assets/b108e30e-bc82-4298-b1bf-8e3477a43717">

**Reply by JasminDreasond** [reply]
this is the results of London without the other countries. I need the Brazil, USA, Mexico in the results too.

**Reply by JasminDreasond** [reply]
I don't know if I should post it on issues. This was to theoretically perform exactly what I commented above. I have a value of a message to receive that already has the values of which Room belongs, but I'm adding another condition that this value also needs to vary between 3 responses of the same string input.

Everything works normally when I put only one extra condition value, but when I try to put multiple ORs, everything fails.
My explanation above is a realistic simulation of what I'm trying to do.

**Reply by JasminDreasond** [reply]
Forget about London, I updated the URL above to get easier because I realized that the example wasn't going to work properly.
Sorry for the incorrect example, I was posting 4:00 am and I was very tired from the work where I found the possible glitch.

**Comment by ujjwalguptaofficial** [comment]
I will have a look and provide resolution ASAP.

**Comment by ujjwalguptaofficial** [comment]
What you want is this - 

```
select({
    from: "Customers",
    where: [{
        city: 'London',
    }, {
        or: {
            country: 'USA'
        },
    }, {
        or: {
            country: 'Mexico'
        }
    }, {
        or: {
            country: 'Brazil'
        }
    }]
});
```

https://ujjwalguptaofficial.github.io/idbstudio/?db=Demo&query=select(%7B%0A%20%20%20%20from%3A%20%22Customers%22%2C%0A%20%20%20%20where%3A%20%5B%7B%0A%20%20%20%20%20%20%20%20city%3A%20'London'%2C%0A%20%20%20%20%7D%2C%20%7B%0A%20%20%20%20%20%20%20%20or%3A%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20country%3A%20'USA'%0A%20%20%20%20%20%20%20%20%7D%2C%0A%20%20%20%20%7D%2C%20%7B%0A%20%20%20%20%20%20%20%20or%3A%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20country%3A%20'Mexico'%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%2C%20%7B%0A%20%20%20%20%20%20%20%20or%3A%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20country%3A%20'Brazil'%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%5D%0A%7D)%3B

**Reply by ujjwalguptaofficial** [reply]
each curly braces represent and and if you want to do `or`, explicityly use `or`

**Reply by JasminDreasond** [reply]
That's the old url. I updated the discussion topic, adding more extra details to help you understand better the subject.

**Reply by JasminDreasond** [reply]
i think you better go on Github to see the answers because i edited some posts to fix my messages.

**Reply by ujjwalguptaofficial** [reply]
@JasminDreasond I have rectified your query in the above link. There was issue in how you were querying. Please check the new one.

**Reply by ujjwalguptaofficial** [reply]
It's different link than yours.

**Comment by ujjwalguptaofficial** [comment]
If you just want all or's condition without the city:'london', then here is equivalent query - 

```
select({
    from: "Customers",
    where: [{
        or: {
            country: 'USA'
        },
    }, {
        or: {
            country: 'Mexico'
        }
    }, {
        or: {
            country: 'Brazil'
        }
    }]
});
```

**Reply by JasminDreasond** [reply]
The test failed. I really received the values of the variety of 3 options, but the conditions of the first "and" were totally ignored, making all the messages of all the rooms to be received at the same time. The "or" conditions needs to respect the first "and" condition.

**Reply by JasminDreasond** [reply]
The where value reference:
```json
[
  {
    "room_id": "...",
    "redaction": false,
    "is_transaction": false
  },
  {
    "or": {
      "member_type": "join"
    }
  },
  {
    "or": {
      "member_type": "leave"
    }
  },
  {
    "or": {
      "member_type": "invite"
    }
  }
]
```

**Reply by ujjwalguptaofficial** [reply, marked as answer]
The query is wrong as per your expectation. If in first query the condition fails and or condition satisfies then data is counted as valid.

You are writing - 

```
if(roomid && redaction && is_transaction) or (member_type=join) or (member_type=leave) or member_type = invite
```

I hope this makes sense.

If you want to return data only when those three column satisfy then you need to use `and` query with member_type value also.

```
[
  {
    "room_id": "...",
    "redaction": false,
    "is_transaction": false,
    "member_type" :{"in":["join","leave","invite"]}
  },
]
```

**Reply by JasminDreasond** [reply]
yep. now really made sense and worked properly. thank you! 🥳
I thought "In" was just to use in value searches inside arrays

i just have one last question now before closing the discussion definitely:
if instead of being "member_type", it's two different columns to use something like "In"?
* Example: the search fail to find the value in member_type then I will look in the member_type_2

**Reply by ujjwalguptaofficial** [reply]
yes you can use any column with in similarly.

**Reply by JasminDreasond** [reply]
that's not it. I'm saying something like it:
```js
if(room_id === value && (member_type === value2 || member_type_2 === value2))
```
